File: impls/reference/wash/metadata.py
"""Command metadata loading and validation."""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import Literal

RECOGNIZED_FIELDS = frozenset(
    {
        "arity",
        "input",
        "output",
        "methods",
        "mime",
        "mutates",
        "parse-mode",
        "stderr",
        "exit",
    }
)

VALID_INPUT = frozenset({"stdin"})
VALID_OUTPUT = frozenset({"stdout"})
VALID_STDERR = frozenset({"discard", "merge"})
VALID_PARSE_MODE = frozenset({"normal", "raw"})
RESERVED_INPUT = frozenset({"file", "none"})
RESERVED_OUTPUT = frozenset({"file"})
RANGE_ARITY_RE = re.compile(r"^\d+\.\.\*?$|^\d+\.\.\d+$")


@dataclass
class ExitMapping:
    explicit: dict[int, int] = field(default_factory=dict)
    wildcard: int | None = None


@dataclass
class CommandMetadata:
    arity: int | Literal["*"] = 0
    input_mode: str = "stdin"
    output_mode: str = "stdout"
    methods: list[str] = field(default_factory=lambda: ["GET"])
    mime: str | None = None
    mutates: bool = False
    parse_mode: str = "normal"
    stderr_mode: str = "discard"
    exit_mapping: ExitMapping = field(default_factory=ExitMapping)
    malformed: bool = False
    malformed_reason: str | None = None

    @property
    def invalid(self) -> bool:
        return self.malformed


def _parse_bool(value: str) -> bool | None:
    if value == "true":
        return True
    if value == "false":
        return False
    return None


def _parse_exit_pairs(tokens: list[str]) -> tuple[ExitMapping | None, str | None]:
    mapping = ExitMapping()
    for token in tokens:
        if "=" not in token:
            return None, f"malformed exit pair: {token!r}"
        code_s, status_s = token.split("=", 1)
        try:
            status = int(status_s)
        except ValueError:
            return None, f"malformed exit status: {status_s!r}"
        if code_s == "*":
            mapping.wildcard = status
        else:
            try:
                code = int(code_s)
            except ValueError:
                return None, f"malformed exit code: {code_s!r}"
            if code < 0:
                return None, f"negative exit code: {code_s!r}"
            mapping.explicit[code] = status
    return mapping, None


def _apply_field(meta: CommandMetadata, name: str, values: list[str]) -> str | None:
    if name == "arity":
        if len(values) != 1:
            return "arity requires exactly one value"
        val = values[0]
        if val == "*":
            meta.arity = "*"
            return None
        if RANGE_ARITY_RE.match(val):
            return f"reserved range arity: {val!r}"
        try:
            n = int(val)
        except ValueError:
            return f"malformed arity: {val!r}"
        if n < 0:
            return f"negative arity: {n}"
        meta.arity = n
        return None

    if name == "input":
        if len(values) != 1:
            return "input requires exactly one value"
        val = values[0]
        if val in RESERVED_INPUT:
            return f"reserved input mode: {val!r}"
        if val not in VALID_INPUT:
            return f"malformed input: {val!r}"
        meta.input_mode = val
        return None

    if name == "output":
        if len(values) != 1:
            return "output requires exactly one value"
        val = values[0]
        if val in RESERVED_OUTPUT:
            return f"reserved output mode: {val!r}"
        if val not in VALID_OUTPUT:
            return f"malformed output: {val!r}"
        meta.output_mode = val
        return None

    if name == "methods":
        if not values:
            return "methods requires at least one value"
        meta.methods = list(values)
        return None

    if name == "mime":
        if len(values) != 1:
            return "mime requires exactly one value"
        val = values[0]
        if "/" not in val or val.strip() != val or not val:
            return f"malformed mime: {val!r}"
        meta.mime = val
        return None

    if name == "mutates":
        if len(values) != 1:
            return "mutates requires exactly one value"
        parsed = _parse_bool(values[0])
        if parsed is None:
            return f"malformed mutates: {values[0]!r}"
        meta.mutates = parsed
        return None

    if name == "parse-mode":
        if len(values) != 1:
            return "parse-mode requires exactly one value"
        val = values[0]
        if val not in VALID_PARSE_MODE:
            return f"malformed parse-mode: {val!r}"
        meta.parse_mode = val
        return None

    if name == "stderr":
        if len(values) != 1:
            return "stderr requires exactly one value"
        val = values[0]
        if val not in VALID_STDERR:
            return f"malformed stderr: {val!r}"
        meta.stderr_mode = val
        return None

    if name == "exit":
        mapping, err = _parse_exit_pairs(values)
        if err:
            return err
        meta.exit_mapping = mapping  # type: ignore[assignment]
        return None

    return None


def load_metadata(root: Path, command_name: str) -> CommandMetadata:
    """Load metadata for a command, applying defaults for missing fields."""
    meta = CommandMetadata()
    meta_path = root / "env" / "meta" / command_name
    if not meta_path.is_file():
        return meta

    fields: dict[str, list[str]] = {}
    try:
        text = meta_path.read_text(encoding="utf-8")
    except OSError:
        meta.malformed = True
        meta.malformed_reason = "failed to read metadata file"
        return meta

    for line in text.splitlines():
        stripped = line.strip()
        if not stripped:
            continue
        if stripped[0] == "#":
            continue
        parts = stripped.split()
        if not parts:
            continue
        field_name = parts[0]
        if field_name not in RECOGNIZED_FIELDS:
            continue
        fields[field_name] = parts[1:]

    for field_name, values in fields.items():
        err = _apply_field(meta, field_name, values)
        if err:
            meta.malformed = True
            meta.malformed_reason = err
            return meta

    if "GET" in meta.methods and meta.mutates:
        meta.malformed = True
        meta.malformed_reason = "GET permitted with mutates true"
        return meta

    return meta


def default_metadata() -> CommandMetadata:
    return CommandMetadata()


def map_exit_status(meta: CommandMetadata, exit_code: int) -> int:
    if exit_code == 0:
        explicit = meta.exit_mapping.explicit.get(0)
        if explicit is not None:
            return explicit
        return 200
    explicit = meta.exit_mapping.explicit.get(exit_code)
    if explicit is not None:
        return explicit
    if meta.exit_mapping.wildcard is not None:
        return meta.exit_mapping.wildcard
    return 400


def method_permitted(meta: CommandMetadata, method: str) -> bool:
    return method in meta.methods


def head_permitted(meta: CommandMetadata) -> bool:
    # GET permitted => HEAD answered; explicit HEAD in methods also permits it.
    return "GET" in meta.methods or "HEAD" in meta.methods

File: impls/dart/lib/metadata/metadata.dart
import 'dart:convert';
import 'dart:io';

const Set<String> recognizedFields = {
  'arity',
  'input',
  'output',
  'methods',
  'mime',
  'mutates',
  'parse-mode',
  'stderr',
  'exit',
};

const Set<String> validInput = {'stdin'};
const Set<String> validOutput = {'stdout'};
const Set<String> validStderr = {'discard', 'merge'};
const Set<String> validParseMode = {'normal', 'raw'};
const Set<String> reservedInput = {'file', 'none'};
const Set<String> reservedOutput = {'file'};

final _rangeArityRe = RegExp(r'^\d+\.\.\*?$|^\d+\.\.\d+$');

class ExitMapping {
  final Map<int, int> explicit;
  final int? wildcard;

  ExitMapping({Map<int, int>? explicit, this.wildcard})
      : explicit = explicit ?? {};
}

class CommandMetadata {
  Object arity; // int or '*'
  String inputMode;
  String outputMode;
  List<String> methods;
  String? mime;
  bool mutates;
  String parseMode;
  String stderrMode;
  ExitMapping exitMapping;
  bool malformed;
  String? malformedReason;

  CommandMetadata({
    this.arity = 0,
    this.inputMode = 'stdin',
    this.outputMode = 'stdout',
    List<String>? methods,
    this.mime,
    this.mutates = false,
    this.parseMode = 'normal',
    this.stderrMode = 'discard',
    ExitMapping? exitMapping,
    this.malformed = false,
    this.malformedReason,
  })  : methods = methods ?? ['GET'],
        exitMapping = exitMapping ?? ExitMapping();

  bool get invalid => malformed;
}

bool? _parseBool(String value) {
  if (value == 'true') return true;
  if (value == 'false') return false;
  return null;
}

(ExitMapping?, String?) _parseExitPairs(List<String> tokens) {
  final explicit = <int, int>{};
  int? wildcard;
  for (final token in tokens) {
    if (!token.contains('=')) {
      return (null, 'malformed exit pair: $token');
    }
    final eqIdx = token.indexOf('=');
    final codeStr = token.substring(0, eqIdx);
    final statusStr = token.substring(eqIdx + 1);
    final status = int.tryParse(statusStr);
    if (status == null) {
      return (null, 'malformed exit status: $statusStr');
    }
    if (codeStr == '*') {
      wildcard = status;
    } else {
      final code = int.tryParse(codeStr);
      if (code == null) {
        return (null, 'malformed exit code: $codeStr');
      }
      if (code < 0) {
        return (null, 'negative exit code: $codeStr');
      }
      explicit[code] = status;
    }
  }
  return (ExitMapping(explicit: explicit, wildcard: wildcard), null);
}

String? _applyField(CommandMetadata meta, String name, List<String> values) {
  if (name == 'arity') {
    if (values.length != 1) return 'arity requires exactly one value';
    final val = values[0];
    if (val == '*') {
      meta.arity = '*';
      return null;
    }
    if (_rangeArityRe.hasMatch(val)) {
      return 'reserved range arity: $val';
    }
    final n = int.tryParse(val);
    if (n == null) return 'malformed arity: $val';
    if (n < 0) return 'negative arity: $n';
    meta.arity = n;
    return null;
  }

  if (name == 'input') {
    if (values.length != 1) return 'input requires exactly one value';
    final val = values[0];
    if (reservedInput.contains(val)) return 'reserved input mode: $val';
    if (!validInput.contains(val)) return 'malformed input: $val';
    meta.inputMode = val;
    return null;
  }

  if (name == 'output') {
    if (values.length != 1) return 'output requires exactly one value';
    final val = values[0];
    if (reservedOutput.contains(val)) return 'reserved output mode: $val';
    if (!validOutput.contains(val)) return 'malformed output: $val';
    meta.outputMode = val;
    return null;
  }

  if (name == 'methods') {
    if (values.isEmpty) return 'methods requires at least one value';
    meta.methods = List<String>.from(values);
    return null;
  }

  if (name == 'mime') {
    if (values.length != 1) return 'mime requires exactly one value';
    final val = values[0];
    if (!val.contains('/') || val.trim() != val || val.isEmpty) {
      return 'malformed mime: $val';
    }
    meta.mime = val;
    return null;
  }

  if (name == 'mutates') {
    if (values.length != 1) return 'mutates requires exactly one value';
    final parsed = _parseBool(values[0]);
    if (parsed == null) return 'malformed mutates: ${values[0]}';
    meta.mutates = parsed;
    return null;
  }

  if (name == 'parse-mode') {
    if (values.length != 1) return 'parse-mode requires exactly one value';
    final val = values[0];
    if (!validParseMode.contains(val)) return 'malformed parse-mode: $val';
    meta.parseMode = val;
    return null;
  }

  if (name == 'stderr') {
    if (values.length != 1) return 'stderr requires exactly one value';
    final val = values[0];
    if (!validStderr.contains(val)) return 'malformed stderr: $val';
    meta.stderrMode = val;
    return null;
  }

  if (name == 'exit') {
    final (mapping, err) = _parseExitPairs(values);
    if (err != null) return err;
    meta.exitMapping = mapping!;
    return null;
  }

  return null;
}

CommandMetadata loadMetadata(String root, String commandName) {
  final meta = CommandMetadata();
  final metaPath =
      '$root${Platform.pathSeparator}env${Platform.pathSeparator}meta${Platform.pathSeparator}$commandName';
  final metaFile = File(metaPath);
  if (!metaFile.existsSync()) return meta;

  String text;
  try {
    text = metaFile.readAsStringSync(encoding: utf8);
  } catch (e) {
    meta.malformed = true;
    meta.malformedReason = 'failed to read metadata file';
    return meta;
  }

  final fields = <String, List<String>>{};
  for (final line in text.split('\n')) {
    final stripped = line.trim();
    if (stripped.isEmpty) continue;
    if (stripped[0] == '#') continue;
    final parts = stripped.split(RegExp(r'\s+'));
    if (parts.isEmpty) continue;
    final fieldName = parts[0];
    if (!recognizedFields.contains(fieldName)) continue;
    fields[fieldName] = parts.sublist(1);
  }

  for (final entry in fields.entries) {
    final err = _applyField(meta, entry.key, entry.value);
    if (err != null) {
      meta.malformed = true;
      meta.malformedReason = err;
      return meta;
    }
  }

  if (meta.methods.contains('GET') && meta.mutates) {
    meta.malformed = true;
    meta.malformedReason = 'GET permitted with mutates true';
    return meta;
  }

  return meta;
}

CommandMetadata defaultMetadata() => CommandMetadata();

int mapExitStatus(CommandMetadata meta, int exitCode) {
  if (exitCode == 0) {
    final explicit = meta.exitMapping.explicit[0];
    if (explicit != null) return explicit;
    return 200;
  }
  final explicit = meta.exitMapping.explicit[exitCode];
  if (explicit != null) return explicit;
  if (meta.exitMapping.wildcard != null) return meta.exitMapping.wildcard!;
  return 400;
}

bool methodPermitted(CommandMetadata meta, String method) {
  return meta.methods.contains(method);
}

bool headPermitted(CommandMetadata meta) {
  return meta.methods.contains('GET') || meta.methods.contains('HEAD');
}
